docs/collection-options-creator.md

```diff
@@ -0,0 +1,667 @@
+  }
+  
+  serialize: {
+    created_at: (date: Date) => Math.floor(date.valueOf() / 1000),  // Date -> timestamp
+    updated_at: (date: Date) => Math.floor(date.valueOf() / 1000),  // Date -> timestamp  
+    metadata?: (obj: object) => JSON.stringify(obj)                 // object -> JSON string
+  }
+}
+
+### 5. Mutation Handler Patterns
+
+There are two distinct patterns for handling mutations in collection options creators:
+
+#### Pattern A: User-Provided Handlers (Electric SQL, Query)
+
+The user provides mutation handlers in the config. Your collection creator passes them through:
+
+```typescript
+interface MyCollectionConfig<TItem extends object> {
+  // ... other config
+  
+  // User provides these handlers
+  onInsert?: InsertMutationFn<TItem>
+  onUpdate?: UpdateMutationFn<TItem>
+  onDelete?: DeleteMutationFn<TItem>
+}
+
+export function myCollectionOptions<TItem extends object>(
+  config: MyCollectionConfig<TItem>
+) {
+  return {
+    // ... other options
+    
+    // Pass through user-provided handlers (possibly with additional logic)
+    onInsert: config.onInsert ? async (params) => {
+      const result = await config.onInsert!(params)
+      // Additional sync coordination logic
+      return result
+    } : undefined
+  }
+}
+```
+
+#### Pattern B: Built-in Handlers (Trailbase, WebSocket)
+
+Your collection creator implements the handlers directly using the sync engine's APIs:
+
+```typescript
+interface MyCollectionConfig<TItem extends object> 
+  extends Omit<CollectionConfig<TItem>, 'onInsert' | 'onUpdate' | 'onDelete'> {
+  // ... sync engine specific config
+  // Note: onInsert/onUpdate/onDelete are NOT in the config
+}
+
+export function myCollectionOptions<TItem extends object>(
+  config: MyCollectionConfig<TItem>
+) {
+  return {
+    // ... other options
+    
+    // Implement handlers using sync engine APIs
+    onInsert: async ({ transaction }) => {
+      const ids = await config.recordApi.createBulk(
+        transaction.mutations.map(m => serialize(m.modified))
+      )
+      await awaitIds(ids)
+      return ids
+    },
+    
+    onUpdate: async ({ transaction }) => {
+      await Promise.all(
+        transaction.mutations.map(m => 
+          config.recordApi.update(m.key, serialize(m.changes))
+        )
+      )
+      await awaitIds(transaction.mutations.map(m => String(m.key)))
+    }
+  }
+}
+```
+
+Choose Pattern A when users need to provide their own APIs, and Pattern B when your sync engine handles writes directly.
+
+## Complete Example: WebSocket Collection
+
+Here's a complete example of a WebSocket-based collection options creator that demonstrates the full round-trip flow:
+
+1. Client sends transaction with all mutations batched together
+2. Server processes the transaction and may modify the data (validation, timestamps, etc.)
+3. Server sends back acknowledgment and the actual processed data
+4. Client waits for this round-trip before dropping optimistic state
+
+```typescript
```

Comment:
Could this be a collection in the lib? If so, could the code be pulled in from an actual file include?

Reply:
No as the backend is entirely custom. Anyone building a custom sync engine would just want to copy this into their codebase as they'd probably inevitably tweak it to meet some particularity of their code base/semantics.